### PR TITLE
Upload songs feature

### PR DESCRIPTION
**Pull request by mohseengs**
**changelog :** 
1. added song model
2. added category model
3. added language model
4. created has_many_belongs_to association between songs and use
5. some modification in User Interface

### CODE REVIEW

File: Musico/app/assets/stylesheets/application.css [hunk 1]
```diff
@@ -13,7 +13,80 @@
...
+    }
+    
+    @media (max-width: 991.98px) {
+    .sidebar {
+    width: 100%;
+    }
+    }
+    .sidebar .active {
+    border-radius: 50px;
+    background-color: var(--accent-extra-light) !important;
```
Comment (mgandhi-gammastack) [inline Musico/app/assets/stylesheets/application.css]: @mohseengs do not use !important

File: Musico/app/assets/stylesheets/authentication.scss
Comment (mgandhi-gammastack) [inline Musico/app/assets/stylesheets/authentication.scss:1]: @mohseengs !important

Comment (mgandhi-gammastack) [reply Musico/app/assets/stylesheets/authentication.scss:1]: for colors https://getbootstrap.com/docs/5.0/customize/color/

File: Musico/app/controllers/songs_controller.rb
Comment (mgandhi-gammastack) [inline Musico/app/controllers/songs_controller.rb:1]: @mohseengs do we need format.json here

File: Musico/app/models/category.rb [hunk 2]
```diff
@@ -0,0 +1,3 @@
+class Category < ApplicationRecord
+    has_many :songs
```
Comment (mgandhi-gammastack) [inline Musico/app/models/category.rb]: use dependent: :destroy for all direct has_many assciations

File: Musico/app/views/layouts/application.html.erb [hunk 3]
```diff
@@ -8,9 +8,19 @@
 
     <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
     <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
+    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-EVSTQN3/azprG1Anm3QDgpJLIm9Nao0Yz1ztcQTwFspd3yD65VohhpuuCOmLASjC" crossorigin="anonymous">
```
Comment (mgandhi-gammastack) [inline Musico/app/views/layouts/application.html.erb]: use bootstrap gem

File: Musico/app/views/songs/index.json.jbuilder [hunk 4]
```diff
@@ -0,0 +1 @@
+json.array! @songs, partial: "songs/song", as: :song
```
Comment (mgandhi-gammastack) [inline Musico/app/views/songs/index.json.jbuilder]: do we need this?

File: Musico/db/seeds.rb [hunk 5]
```diff
@@ -5,3 +5,23 @@
...
+Category.create(:name => "Accustic")
+Category.create(:name => "Other")
+
+
+Language.create(:name => "Hindi")
+Language.create(:name => "English")
+Language.create(:name => "Punjabi")
+
+User.create(:name => "Mohseen",:email => "mohseentech@gmail.com",:password => "12345",:dob => "12-03-2000")
+User.create(:name => "sumit",:email => "sumit@gmail.com",:password => "12345",:dob => "12-03-2000")
```
Comment (mgandhi-gammastack) [inline Musico/db/seeds.rb]: use loops and arrays

File: Musico/test/controllers/songs_controller_test.rb
Comment (mgandhi-gammastack) [inline Musico/test/controllers/songs_controller_test.rb:1]: do we need test folder?

File: Musico/app/controllers/songs_controller.rb [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    @categories = Category.all
+    @languages = Language.all
+  end
+
+  # GET /songs/1/edit
+  def edit
+  end
+
+  def create
+    @song = Song.new(song_params)
```
Comment (mgandhi-gammastack) [inline Musico/app/controllers/songs_controller.rb]: song should belongs to a user right?

File: Musico/db/seeds.rb (hunk 5 above)
Comment (mgandhi-gammastack) [reply Musico/db/seeds.rb]: Also use bang methods like create!

File: Musico/config/routes.rb [hunk 7]
```diff
@@ -1,4 +1,5 @@
 Rails.application.routes.draw do
+  resources :songs
```
Comment (mgandhi-gammastack) [inline Musico/config/routes.rb]: check order

File: Musico/app/models/song.rb [hunk 8]
```diff
@@ -0,0 +1,5 @@
+class Song < ApplicationRecord
+  has_and_belongs_to_many :users
```
Comment (mgandhi-gammastack) [inline Musico/app/models/song.rb]: dependent condition

File: Musico/app/controllers/application_controller.rb [hunk 9]
```diff
@@ -1,2 +1,12 @@
 class ApplicationController < ActionController::Base
+    include ApplicationHelper
```
Comment (mgandhi-gammastack) [inline Musico/app/controllers/application_controller.rb]: why we are including ApplicationHelper here?, helpers are used for different purpose. create a concern if you want to.

Comment (mgandhi-gammastack) [reply Musico/app/controllers/application_controller.rb]: And if you decide to create a concern also move this check_login method also there

File: Musico/app/controllers/authentication_controller.rb [hunk 10]
```diff
@@ -0,0 +1,31 @@
+class AuthenticationController < ApplicationController
+  skip_before_action :check_login
+  layout "authentication"
```
Comment (mgandhi-gammastack) [inline Musico/app/controllers/authentication_controller.rb]: I hope that this will be changed as per our discussion.

**Comment by mgandhi-gammastack** [comment]
recreate this once dependent branch is merged to main

File: Musico/app/controllers/songs_controller.rb
Comment (mohseengs) [reply Musico/app/controllers/songs_controller.rb:1]: done

File: Musico/app/models/user.rb [hunk 11]
```diff
@@ -3,4 +3,7 @@ class User < ApplicationRecord
     EMAIL_REGEX = /\A[^@\s]+@[^@\s]+\z/
     validates :email, presence: true, uniqueness: true, format: { with: EMAIL_REGEX, message: 'Invalid email' }
     validates :name, :password, :dob, presence: :true
+    has_and_belongs_to_many :songs
```
Comment (mohseengs) [inline Musico/app/models/user.rb]: dependent condition

File: Musico/Gemfile [hunk 12]
```diff
@@ -54,4 +56,4 @@ end
 
 # Windows does not include zoneinfo files, so bundle the tzinfo-data gem
 gem 'tzinfo-data', platforms: [:mingw, :mswin, :x64_mingw, :jruby]
-gem 'awesome_print'
\ No newline at end of file
+gem 'awesome_print'
```
Comment (mgandhi-gammastack) [inline Musico/Gemfile]: redundant gem, already added above

File: Musico/app/controllers/songs_controller.rb [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+  # GET /songs/1/edit
+  def edit
+  end
+
+  def create
+      @song = Song.new(song_params)
+      if @song.save
+        @song.user_ids = params[:song][:users].map{ |id| id.to_i }
+        @song.users << current_user
```
Comment (mgandhi-gammastack) [inline Musico/app/controllers/songs_controller.rb]: Do this using `song_params` only

File: Musico/app/controllers/songs_controller.rb [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+  # DELETE /songs/1 or /songs/1.json
+  def destroy
+      @song.destroy!
+
+      redirect_to songs_url, notice: "Song was successfully destroyed."
+  end
+
+  def search_user
+      query = params[:query]
+      @users = User.where("name LIKE ? and id <> ?", "%" + query + "%" , current_user.id )
```
Comment (mgandhi-gammastack) [inline Musico/app/controllers/songs_controller.rb]: convert this into a scope in user, where ever possible do not use raw sql. like here for id, you can use `where.not(id: user_id)`

File: Musico/app/views/songs/_song_users.html.erb [hunk 15]
```diff
@@ -0,0 +1,19 @@
+<% @users.each do |user| %>
+    <div class="checkbox_div">
+        <%= check_box_tag "song[users][]" , user.id, false, label: user.name, id: :search_user_checkbox %>
```
Comment (mgandhi-gammastack) [inline Musico/app/views/songs/_song_users.html.erb]: You should not use same id for multiple elements in the same page

File: Musico/Gemfile (hunk 12 above)
Comment (mohseengs) [reply Musico/Gemfile]: done

File: Musico/app/controllers/songs_controller.rb (hunk 13 above)
Comment (mohseengs) [reply Musico/app/controllers/songs_controller.rb]: done

File: Musico/app/controllers/songs_controller.rb (hunk 14 above)
Comment (mohseengs) [reply Musico/app/controllers/songs_controller.rb]: done

File: Musico/app/views/songs/_song_users.html.erb (hunk 15 above)
Comment (mohseengs) [reply Musico/app/views/songs/_song_users.html.erb]: done

**Pull request closed without merging**